crates/core/src/rlp/decode.rs

```diff
@@ -230,20 +235,94 @@ impl<T: RLPDecode> RLPDecode for Vec<T> {
     }
 }
 
-fn decode_list(data: &[u8]) -> Result<&[u8], RLPDecodeError> {
+impl<T1: RLPDecode, T2: RLPDecode> RLPDecode for (T1, T2) {
+    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
+        if rlp.is_empty() {
+            return Err(RLPDecodeError::InvalidLength);
+        }
+        let (is_list, payload, _) = decode_rlp_item(rlp)?;
+        if !is_list {
+            return Err(RLPDecodeError::MalformedData);
+        }
+
+        let (_, first, rest) = decode_rlp_item(payload)?;
+        let first = if first.is_empty() {
+            T1::decode(&[RLP_EMPTY_LIST])?
+        } else {
+            T1::decode(payload)?
+        };
+        let second = T2::decode(rest)?;
+        Ok((first, second))
+    }
+}
+
+impl<T1: RLPDecode, T2: RLPDecode, T3: RLPDecode> RLPDecode for (T1, T2, T3) {
+    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
+        if rlp.is_empty() {
+            return Err(RLPDecodeError::InvalidLength);
+        }
+        let (is_list, payload, _) = decode_rlp_item(rlp)?;
+        if !is_list {
+            return Err(RLPDecodeError::MalformedData);
+        }
+
+        let (_, first, first_rest) = decode_rlp_item(payload)?;
+        let first_decoded = if first.is_empty() {
+            T1::decode(&[RLP_EMPTY_LIST])?
+        } else {
+            T1::decode(payload)?
+        };
+
+        let (_, second, second_rest) = decode_rlp_item(first_rest)?;
+        let second_decoded = if second.is_empty() {
+            T2::decode(&[RLP_EMPTY_LIST])?
+        } else {
+            T2::decode(first_rest)?
+        };
+        let third_decoded = T3::decode(second_rest)?;
+
+        Ok((first_decoded, second_decoded, third_decoded))
+    }
+}
+
+fn decode_rlp_item(data: &[u8]) -> Result<(bool, &[u8], &[u8]), RLPDecodeError> {
```

Comment:
We should replace the `bool` with an enum, either wrapping the current item slice:
```rust
enum RLPItem<'a> {
    Int(u8),
    String(&'a [u8]),
    List(&'a [u8]),
}
// -> Result<(RLPItem, &[u8]), RLPDecodeError>
```
or not:
```rust
enum RLPItem {
    String,
    List,
}
// -> Result<(RLPItem, &[u8], &[u8]), RLPDecodeError>
```